numba/parfor.py

```diff
@@ -3162,14 +3178,568 @@ def __init__(self, blocks):
+    for k,v in blocks.items():
+        # Create a new block for aliased code.
+        new_orig_block = ir.Block(v.scope, v.loc)
+        # Create a new block for non-aliased code.
+        block_copy = ir.Block(v.scope, v.loc)
+        # For each statement in the current block...
+        for stmt in v.body:
+            # We most use the original code for the aliased case.  The only thing we have to change
+            # is the branch and jump labels.
+            if isinstance(stmt, ir.Jump):
+                new_orig_block.body.append(ir.Jump(stmt.target + alias_offset, stmt.loc))
+            elif isinstance(stmt, ir.Branch):
+                new_orig_block.body.append(ir.Branch(stmt.cond, stmt.truebr + alias_offset, stmt.falsebr + alias_offset, stmt.loc))
+            else:
+                # Add the original statement unmodified to the new aliased code block.
+                new_orig_block.body.append(stmt)
+
+            # Add the statement to the non-aliased code block.  Have to recursively replace all variables
+            # and update jump and branch labels.
+            block_copy.body.append(duplicate_stmt(stmt, typemap, calltypes, dup_typemap, dup_calltypes, noalias_offset))
+        new_blocks[k + alias_offset] = new_orig_block
+        new_blocks[k + noalias_offset] = block_copy
+
+    typemap.update(dup_typemap)
+    calltypes.update(dup_calltypes)
+
+    return new_blocks
+
+def generate_aliasing_variants(func_ir, typemap, calltypes, array_analysis, typingctx, flags):
+    """
+    If an array argument is written to then it may alias any of the other array arguments.
+    If so, check for aliasing dynamically and create two variants, one for aliasing and one without.
+    """
+    if flags.noalias:
+        return False
+
+    blocks = func_ir.blocks
+    alias_map, arg_aliases = find_potential_aliases(blocks, func_ir.arg_names, typemap, func_ir, None, None, flags.noalias)
+    if config.DEBUG_ARRAY_OPT >= 1:
+        print("generate_aliasing_variants")
+        print("alias_map:", alias_map)
+        print("arg_aliases:", arg_aliases)
+
+    array_args = {}
+    for arg_index, an in enumerate(func_ir.arg_names):
+        if isinstance(typemap[an], types.npytypes.Array):
+            array_args[an] = arg_index
+    if len(array_args) < 2:
+        return False
+
+    args_written_to = False
+    for block in blocks.values():
+        args_written_to = find_args_written_to(block, arg_aliases)
+        if args_written_to:
+            break
+    if config.DEBUG_ARRAY_OPT >= 1:
+        print("args_written_to:", args_written_to)
+
+    if args_written_to:
+        """ If we found some input array or some alias thereof written into by this function
+            then generate code to check potential aliasing at runtime.
+        """
+        dprint_func_ir(func_ir, "before alias duplication")
+        func_ir.blocks = duplicate_code(blocks, typemap, calltypes, func_ir.arg_names, typingctx, array_args)
```

Comment:
This function needs to guard against failing in `duplicate_code`.

For example:

```python
import numpy as np
from numba import njit


@njit(parallel=True)
def foo(a, b, out):
    out[...] = a + b


a = np.random.random(10)
b = np.random.random(10)
out = np.empty_like(a)
foo(a, b, out)
print(out)
```